### ISSUE TITLE
Is there an option to run tf2onnx without optimizing ONNX model?

### ISSUE THREAD

**Issue by wm2012011492** [onnx/tensorflow-onnx#1027]
**Describe the bug**
tf2onnx.optimizer might change the graph unexpectedly, which causes trouble for latter onnx2trt convertion. 

**System information**
- Linux Ubuntu 18.04:
- Tensorflow Version: 1.15
- Python version: 3.6

**Expected behavior**
tf2onnx provides an option to close optimization step during conversion.

**Comment by guschmue** [comment]
currently not since the graph would look horrible (like lots of transposes, cases and identities).
But the optimizer should not do anything that violates the onnx spec.

If this is a one time need you could comment out this line:
https://github.com/onnx/tensorflow-onnx/blob/master/tf2onnx/convert.py#L157

Else we can add a command line option.

**Comment by TomWildenhain-Microsoft** [comment]
What problems in onnx2trt conversion are you facing, and how is the optimizer causing an issue?

**Comment by wm2012011492** [comment]
In my case, tf2onnx optimizer will not cause any accuracy issue but it changes the graph like combining two weights which are used by two Add op. When it comes to refitting a TRT engine, we need the onnx op or weight name, which has been changed by tf2onnx optimzier. So it might cause trouble in TRT refitting.

**Comment by jay-karan** [comment]
Hi @guschmue @TomWildenhain-Microsoft, 

1. May I know how optimizers defined in [tf2onnx](https://github.com/onnx/tensorflow-onnx/tree/master/tf2onnx/optimizer) different from [onnx.optimize](https://github.com/onnx/optimizer/tree/master/onnxoptimizer) and 
[onnxconverter-common.optimizer](https://github.com/microsoft/onnxconverter-common/tree/master/onnxconverter_common)

2. Does tf2onnx optimizer optimizes the ONNX graph obtained after conversion or the tensorflow graph before converting?

**Comment by guschmue** [comment]
onnx.optimize is really old and insufficient, and we think its not reliable.
onnxconverter-common.optimizer is similar to what tf2onnx does but the optimizers in tf2onnx handle a little more. If we'd start tf2onnx today we'd most likely use onnxconverter-common.optimizer.

We optimize few things before conversion (ie. constant folding) since it helps the conversion but most of what we call optimization we run after conversion.

**Comment by jay-karan** [comment]
Thank you.
- Can we still try to use onnxconverter-common.optimzer on tf graphs while converting?
- Do we need to explicitly make calls to optimize few things before and after conversions?
- Does the use of tf2onnx's optimizer result in considerable change in performance of size, accuracy and inference speed?
- How much is the scope of adding custom optimizations to the existing set of optimizers?

**Comment by guschmue** [comment]
Yes, you could run onnxconverter-common.optimzer after tf2onnx but we think you should not see gains.
The optimizer in tf2onnx runs after the conversion so you need to explicitly call it. But we are thinking to release a new api that will combine those calls.
perf and size will change considerable. Depends on the model how much. For example if the tf model comes with nhwc the model will be horrible slow without optimizer.
The optimizer is not aware of custom ops since we have not had need for it (yet).
There is overlap between tf2onnx and onnxconverter-common.

**Comment by wm2012011492** [comment]
Could I only turn off the --fold-constant optimization? The case is that I need to tile many weights by batch_size, like tf.tile(tf_weights, [batch_size]). Then tf2onnx will do constant_folding for these tile layers. The consequence is I get a super large .onnx file since the stored weights become batch_size times larger than it should be. When batch_size is 256, it even exceeds the maximum protobuf size of 2GB.

**Comment by TomWildenhain-Microsoft** [comment]
This is a problem I thought we might run into at some point.... The constant folding here likely isn't actually from the optimizers, it is from the folding we do before conversion.  Can you attach a copy of your model/code?  We might want to disable folding if the resulting tensor is substantially larger than the inputs to the op.  A temporary workaround for you @wm2012011492 might be converting the model with the batch_size left unspecified.  Then folding won't be able to happen on that node.

**Comment by wm2012011492** [comment]
I can attach a simple model for reproduction later. The temporary workaround seems unacceptable in this case since I'm not allowed to modify the TF source code. How could I turn off --fold-constant in tile layers? I can afford to re-build tf2onnx library.

**Comment by jay-karan** [comment]
Thank you @guschmue . One thing is still ambiguous.

- When we are converting tensorflow graph to onnx graph, the corresponding operators (tf ops are <-> onnx ops) are being replaced right. Then, why is there a need for optimization?

**Comment by guschmue** [comment]
@jay-karan there are quite a few things we can optimize, for example nhwc->nchw transposes, fusing padding into conv, fusing batchnorm into conv, replacing mul/add with a gem, fusing back to back operations, removing duplicate constants ... depending on the model the gains are huge.

@wm2012011492 the --fold-constants thing is not supported for some time and it was only there to deal with some issues in older tensorflow version. Internally we depend on constant folding to deal with dynamic inputs in tensorflow that are attributes in onnx. There is also some pattern matching we do that depends on constant folding.
If I understand your problem correctly: you calculate some constant on the fly and if we fold them the size of the onnx file grows a lot. It should not be to hard to break the constant folding, ie something like add(x, tf.zeros([0]) at the root of your calculation might already break it. If that doesn't do the trick we can help you to find something that will break constant folding.

**Comment by wm2012011492** [comment]
You are right that I could use tf.tile(tf_weights, [tf.shape(input)[0]]) to avoid constant folding. But just like I said before, it's not preferred to modify the tf source code in my case. So I would rather modify tf2onnx and rebuilt this library. Is there any minimum change to turn-off constant folding tile layers?

**Comment by TomWildenhain-Microsoft** [comment]
@wm2012011492 I need to know which part of the codebase is actually doing the folding in this case.  There are 3 places it could be happening.  During conversion there should be a line **converted x values for constant folding**.  What is the value of x here?  Can you post the complete converter output?

**Comment by TomWildenhain-Microsoft** [comment]
@jay-karan To add on to what @guschmue said, part of the justification for including the optimizers in the tf2onnx repo (despite running them after converting the graph to onnx) is that we are often optimizing away inefficiencies introduced by the converter itself.  For example, an onnx op might output a different type than the corresponding TF type so we add a cast.  But if the original graph already has a cast after the op, an additional cast is not necessary.  Rather than checking this at the time of conversion in every op that inserts a cast, it is easier to just add the cast in always and then optimize it out later with the back_to_back optimizer if possible.  Ops that take transposed inputs behave similarly.  We throw in a transpose first and then try to optimize it out later.  Full list of optimizers:

```
const_fold_optimizer
identity_optimizer
merge_duplicated_nodes_optimizer
transpose_optimizer
loop_optimizer
back_to_back_optimizer
upsample_optimizer
```

Let me know if you want more explanation for any of them.

**Comment by wm2012011492** [comment]
```
INFO - tf2onnx.tf_utils: Computed 4 values for constant folding
INFO - tf2onnx.tfonnx: folding node using tf type=Tile, name="weight_tile"
```
Is this what you need?

**Comment by TomWildenhain-Microsoft** [comment]
@wm2012011492 Thanks, can you try converting again with the latest tf2onnx from git?   #1201 should fix your issue.

`pip uninstall tf2onnx`
`pip install git+https://github.com/onnx/tensorflow-onnx`

**Comment by jay-karan** [comment]
> @jay-karan To add on to what @guschmue said, part of the justification for including the optimizers in the tf2onnx repo (despite running them after converting the graph to onnx) is that we are often optimizing away inefficiencies introduced by the converter itself. For example, an onnx op might output a different type than the corresponding TF type so we add a cast. But if the original graph already has a cast after the op, an additional cast is not necessary. Rather than checking this at the time of conversion in every op that inserts a cast, it is easier to just add the cast in always and then optimize it out later with the back_to_back optimizer if possible. Ops that take transposed inputs behave similarly. We throw in a transpose first and then try to optimize it out later. Full list of optimizers:
> 
> ```
> const_fold_optimizer
> identity_optimizer
> merge_duplicated_nodes_optimizer
> transpose_optimizer
> loop_optimizer
> back_to_back_optimizer
> upsample_optimizer
> ```
> 
> Let me know if you want more explanation for any of them.

Thank you @TomWildenhain-Microsoft 

- So, may I conclude that the optimizations defined are only removing the un-necessary operators produced after conversion of tf2onnx? 
- How is the performance improved? Will there be any marginal change in inference times on vision based tensorflow models after converting to onnx?
- May I also ask if there are any possibilities on further accelerating the inference of models by any custom operators or custom optimizations in onnx?

**Comment by TomWildenhain-Microsoft** [comment]
Yes, the optimizers are primarily targeted at removing inefficiencies added during conversion.  Though they occasionally may remove problems with the original TF graph, they are not the primary cause of speedup in most cases.  ONNX is generally faster than TF not because the graph is better but because ORT is just faster than the TF runtime for many ops.  ORT has multiple execution providers for each op that use hardware optimizations depending on the machine (GPU, AVX2 instructions, etc.)

The amount of speedup depends greatly on the model, but I would expect the model to be about 2 times faster with ONNX.  We recently downloaded a bunch of models from TF hub and compared the ONNX vs TF inference times and found the median speedup to be 2.9 times faster.  84% of models were as fast or faster using ORT vs TF.  70% were at least 1.5 times as fast.

**Comment by guschmue** [comment]
I'd say the optimizers do mostly removing, fusing and replacing. Example for the later is if we replace a add/mul with a gem.
We optimize before conversion (mostly via tensorflow/grappler and some extra constant folding) - this we depend on.
We optimize after conversion. Those are purely for perf and size. One could skip this. The impact depends a lot on the model but it is significant. If you have a tensorflow model that is channel last not running the optimizers is going to be very bad for perf.

**Comment by TomWildenhain-Microsoft** [comment]
@wm2012011492 for your purposes of keeping the names the same, as you have noticed, tf2onnx does not guarantee names of the onnx graph match those in the TF graph (except for inputs/outputs).  Disabling the optimizers might be sufficient in your case to keep the names the same.  To disable some/all of them, remove lines from /tf2onnx/optimizer/__init__.py.
